PAL-1094-update-documentation: root README.md (#75)
gchq/Palisade-clients: README.md

Windows is not an explicitly supported environment, although where possible Palisade has been made compatible.  
For Windows developer environments, we recommend setting up [WSL](https://docs.microsoft.com/en-us/windows/wsl/).

## Getting started

> For an overview of Palisade, start with root documentation [Palisade README](https://github.com/gchq/Palisade#readme).

Reviewer: Would something like this read better?
```suggestion
This is not the starting point for understanding Palisade.
This should be done from within the main [Palisade repo](https://github.com/gchq/Palisade)
```

Author: I went for another tack and expanded on what is at the other end of the link "For an overview of Palisade, start with the Palisade introduction and the accompanying guides: QuickStart Guide; and Developer Guide:"

# Palisade Clients

## Client Implementations
Implementation of a client will need to send requests into Palisade and process the response.  
The workflow for this is expected to follow the pattern of first sending in a request for data.  

> This will return a response consisting of a token (a unique id) that can then be used a second request to retrieve the processed data.

Reviewer 2: ```suggestion
This will return a response consisting of a token (a unique id) that can then be used in a second request to retrieve the processed data.
```

> Client implementation will need to conform to this workflow, but the technology that is used and how the data is to be processed has by design meant to be left open for the implementations of the Palisade service.

Reviewer 2: 'left open for the implementations of the Palisade services.' As I was reading this I thought it was talking about the clients being written for the different processing technologies. Which meant that the Palisade services part threw me off here. The technology used for the clients would be dependent on what technology the user intends to use for follow on processing of the data.

Author: I don't even understand it.  I'll try again.

> The following examples demonstrate the use of different kinds of clients that can operate with the Palisade service:

Reviewer 2: Can we have some details of what each client is and therefore what order should people try to understand these clients. Would also be worth pointing out here the hierarchy of these clients. E.g that the akka client uses the java client, etc

## Getting started
For an overview of Palisade, start with the Palisade introduction and the accompanying guides: QuickStart Guide; and Developer Guide: [Palisade README](https://github.com/gchq/Palisade).

# Palisade Clients

## Client Implementations
Implementation of a client will need to send requests into Palisade and process the response.  
The workflow for this is expected to follow the pattern of first sending in a request to register access for data.  
This will return a response consisting of a token (a unique id) that can then be used in a second request to retrieve the processed data.
An implementation of a Client will need to conform to this workflow, but the technology that is used can be left to best suit the intended use of the application.

The following examples demonstrate the use of different kinds of clients that can operate with the Palisade service:
* [Java Client](client-java/README.md)
  Java based RESTFul client using an API similar to the JDBC  
* [Shell Client](client-shell/README.md)
  Command Line Interface(CLI) client which provides interactive operations.
* [Fuse Client](client-fuse/README.md)

>   Filesystem in Userspace (FUSE) client which uses a CLI for interactive operations and stores the results in a FUSE filesystem.  

Reviewer 2: mention that this is an extension on the java client

Author: Added the one sentence "  Client is an extension of the Java Client.".

* [Akka Client](client-akka/README.md)

>   Akka client which provides access an interface using both Java and Akka types.

Reviewer 2: mention that this is also an extension of the java client

Author: Added the one sentence " Client is an extension of the Java Client.".

Reviewer: ```suggestion
  Akka client which provides access to an interface using both Java and Akka types.
```

* [S3 Client](client-s3/README.md)

>   Akka client which provides access an interface using both Java and Akka types and data is prepared and then stored on an Amazon Web Service (AWS) Simple Cloud Storage (S3) bucket from where is it retrieved.

Reviewer 2: am not sure this statement is correct. This client is a microservice that exposes an S3 API REST endpoint which technologies that can normally read data from S3 compliant API's can then route their S3 commands via this microservice to seamlessly get the data access protections palisade provides

Author: Took another stab at it.  It now reads "Akka Microservice which provides an endpoint for accessing resources stored in an Amazon Web Service (AWS) Simple Cloud Storage (S3) bucket.  
  The example uses [Apache Spark](https://sparkjava.com/) for RESTful queries with the microservice."

Windows is not an explicitly supported environment, although where possible Palisade has been made compatible.  
For Windows developer environments, we recommend setting up [WSL](https://docs.microsoft.com/en-us/windows/wsl/).

## Getting started

> For an overview of Palisade, start with the Palisade introduction and the accompanying guides: QuickStart Guide; and Developer Guide: [Palisade README](https://github.com/gchq/Palisade).

Reviewer 3: Direct links to both of these?

Reviewer 2: I think it is better with just the one link, the others are just examples of the types of documents you can find there

Author: left is a is.  This it make more sense this way.  Point them to the starting point

# Palisade Clients

## Client Implementations
The Client implementations will need to send requests into Palisade and process the response.  
The workflow for this is expected to follow the pattern of first sending in a request to register access for data.  

> This will return a response consisting of a token (a unique id) that can then be used in a second request to retrieve the actual data.

Reviewer 3: Second request gathers resources, third gets data

Author: Re-worded along these lines

> Implementation of a client will need to send requests into Palisade and process the response.  
> The workflow for this is expected to follow the pattern of first sending in a request to register access for data.  
> This will return a response consisting of a token (a unique id) that can then be used in a second request to retrieve the processed data.
> An implementation of a Client will need to conform to this workflow, but the technology that is used can be left to best suit the intended use of the application.

Reviewer: ```suggestion
The Client implementations will need to send requests into Palisade and process the response.  
The workflow for this is expected to follow the pattern of first sending in a request to register access for data.  
This will return a response consisting of a token (a unique id) that can then be used in a second request to retrieve the actual data.
An implementation of a Client will need to conform to this workflow, but the technology that is used will be best suited to the intended use of the application.
```

The following examples demonstrate the use of different kinds of clients that can operate with the Palisade service:
* [Java Client](client-java/README.md)
  Java based RESTFul client using an API similar to the JDBC  
* [Shell Client](client-shell/README.md)
  Command Line Interface(CLI) client which provides interactive operations.
* [Fuse Client](client-fuse/README.md)
  Filesystem in Userspace (FUSE) client which uses a CLI for interactive operations and stores the results in a FUSE filesystem.

>   Client is an extension of the Java Client.

Reviewer: ```suggestion
  This client is an extension of the Java Client.
```

* [Akka Client](client-akka/README.md)
  Akka client which provides access an interface using both Java and Akka types.

>   Client is an extension of the Java Client.

Reviewer: ```suggestion
  This client is an extension of the Java Client.
```

* [S3 Client](client-s3/README.md)
  Akka Microservice which provides an endpoint for accessing resources stored in an Amazon Web Service (AWS) Simple Cloud Storage (S3) bucket.  

>   The example uses [Apache Spark](https://sparkjava.com/) for RESTful queries with the microservice. 

Reviewer 3: Prefer https://spark.apache.org/

Author: done

```
Now you can finally build the repository by running: 
```bash

> mvn clean install

Reviewer: Not sure the clean phase is required as this is likely to be a fresh clone of the code

Author: It is not necessary, but I think that it is a good practice.  It will set the user-up the command prompt with the inevitable variant when there is a need to re-do the build

Reviewer 2: I dont mind having the clean in here and it is good practice
